### PR TITLE
Implemented IsSubType Module Successfully

### PR DESCRIPTION
**Pull request by Smit1603**
- This pull request will resolve issue #18 
- I have implemented IsSubType Module and ran it succesfully
- Kindly review it and please let me know whether any improvement or changes are required or not
- @vgvassilev and @sudo-panda

### CODE REVIEW

File: lib/Interpreter/InterOp.cpp [hunk 1]
```diff
@@ -2184,5 +2184,40 @@ namespace InterOp {
...
+
+  bool IsSubType(TCppType_t subclass, TCppType_t superclass)
+  {
+    auto derivedType = clang::QualType::getFromOpaquePtr(subclass);
+    auto baseType = clang::QualType::getFromOpaquePtr(superclass);
+
+    if (derivedType.isNull() || baseType.isNull())
+    {
+      return false;
+    }
```
Comment (vgvassilev) [inline lib/Interpreter/InterOp.cpp:2196]:
```suggestion
    if (derivedType.isNull() || baseType.isNull())
      return false;
```

File: lib/Interpreter/InterOp.cpp [hunk 2]
```diff
@@ -2184,5 +2184,40 @@ namespace InterOp {
...
+
+    auto derivedDecl = derivedType->getAsCXXRecordDecl();
+    auto baseDecl = baseType->getAsCXXRecordDecl();
+
+    if (!derivedDecl || !baseDecl)
+    {
+      return false;
+    }
```
Comment (vgvassilev) [inline lib/Interpreter/InterOp.cpp:2204]:
```suggestion
    if (!derivedDecl || !baseDecl)
      return false;
```

File: lib/Interpreter/InterOp.cpp [hunk 3]
```diff
@@ -2184,5 +2184,40 @@ namespace InterOp {
...
+
+    if (derivedDecl == baseDecl)
+    {
+      return true;
+    }
```
Comment (vgvassilev) [inline lib/Interpreter/InterOp.cpp:2209]:
```suggestion
    if (derivedDecl == baseDecl)
      return true;
```

File: lib/Interpreter/InterOp.cpp [hunk 4]
```diff
@@ -2184,5 +2184,40 @@ namespace InterOp {
...
+
+    for (auto it = derivedDecl->bases_begin(); it != derivedDecl->bases_end(); it++)
+    {
+      auto baseRecordType = it->getType()->getAs<clang::RecordType>();
+      if (baseRecordType && baseRecordType->getDecl() == baseDecl)
+      {
+        return true;
+      }
```
Comment (vgvassilev) [inline lib/Interpreter/InterOp.cpp:2217]:
```suggestion
      if (baseRecordType && baseRecordType->getDecl() == baseDecl)
        return true;
```

**Comment by Smit1603** [comment]
- @vgvassilev I have made changes as mentioned by you 
- Kindly review and pls let me know whether any more changes/improvements are required or not

File: lib/Interpreter/InterOp.cpp [hunk 5]
```diff
@@ -2184,5 +2184,31 @@ namespace InterOp {
     }
     return dims;
   }
+
+  bool IsSubType(TCppType_t subclass, TCppType_t superclass)
+  {
+    auto derivedType = clang::QualType::getFromOpaquePtr(subclass);
+    auto baseType = clang::QualType::getFromOpaquePtr(superclass);
```
Comment (sudo-panda) [inline lib/Interpreter/InterOp.cpp:2191]:
```suggestion
    auto derivedType = clang::QualType::getFromOpaquePtr(derived);
    auto baseType = clang::QualType::getFromOpaquePtr(base);
```

File: unittests/InterOp/TypeReflectionTest.cpp [hunk 6]
```diff
@@ -346,9 +346,9 @@ TEST(TypeReflectionTest, DISABLED_IsSubType) {
   InterOp::TCppType_t type_B = InterOp::GetVariableType(Decls[4]);
   InterOp::TCppType_t type_C = InterOp::GetVariableType(Decls[5]);
 
-  // EXPECT_TRUE(InterOp::IsSubType(type_B, type_A));
-  // EXPECT_FALSE(InterOp::IsSubType(type_A, type_B));
-  // EXPECT_FALSE(InterOp::IsSubType(type_C, type_A));
+  EXPECT_TRUE(InterOp::IsSubType(type_B, type_A));
+  EXPECT_FALSE(InterOp::IsSubType(type_A, type_B));
+  EXPECT_FALSE(InterOp::IsSubType(type_C, type_A));
```
Comment (sudo-panda) [inline unittests/InterOp/TypeReflectionTest.cpp:351]: Can you add more test cases?

**Review by sudo-panda — CHANGES_REQUESTED** [review]

**Comment by Smit1603** [comment]
- @sudo-panda I have changed variables name as mentioned by you for consistency purpose
- With regards to extra test cases I tried to perform multilevel inheritance and tried above implemented module and found that test gives incorrect output 
- So I am currently resolving that issue and will commit everything after resolving above issues 
- Just making sure and informing you whether I am going in correct direction???

**Comment by Smit1603** [comment]
- @sudo-panda as mentioned above
- I have added few more test cases which can check for various types of inheritance
- I have also changed variable names for consistency purpose
- I hope this modified code will meet every test cases and its requirements
- Pls review and let me know whether any further changes/improvements are required or not

File: lib/Interpreter/InterOp.cpp [hunk 7]
```diff
@@ -2184,5 +2184,35 @@ namespace InterOp {
     }
     return dims;
   }
+
+  bool IsSubType(TCppType_t derived, TCppType_t base)
+  {
+    auto derivedType = clang::QualType::getFromOpaquePtr(derived);
+    auto baseType = clang::QualType::getFromOpaquePtr(base);
+
```
Comment (vgvassilev) [inline lib/Interpreter/InterOp.cpp:2192]: I think we could implement that using `Sema::isDerivedFrom`: https://clang.llvm.org/doxygen/classclang_1_1Sema.html#ae9fa2ea2b4ce733c9f50241d25c45132

Comment (vgvassilev) [reply lib/Interpreter/InterOp.cpp:2192]: @sudo-panda, what should that function return if we ask a nested type if it is a subtype?

Comment (sudo-panda) [reply lib/Interpreter/InterOp.cpp:2192]: The input can be assumed to be a `RecordType`

Comment (sudo-panda) [reply lib/Interpreter/InterOp.cpp:2192]: Yeah I agree we are looking for something like "IsTypeDerivedFrom". 

I didn't know that inner classes are also referred to as subtypes.

Comment (vgvassilev) [reply lib/Interpreter/InterOp.cpp:2192]: See my initial comment, this is what we need to do, and change the interface name as @sudo-panda suggested.

**Comment by Smit1603** [comment]
- @vgvassilev and @sudo-panda pls review and let me know whether any further improvements/changes are required or not

File: lib/Interpreter/InterOp.cpp (hunk 7 above)
Comment (vgvassilev) [reply lib/Interpreter/InterOp.cpp:2192]: @Smit1603 this is not yet done

File: lib/Interpreter/InterOp.cpp [hunk 8]
```diff
@@ -2184,5 +2184,25 @@ namespace InterOp {
...
+    auto baseDecl = baseType->getAsCXXRecordDecl();
+
+    if (!derivedDecl || !baseDecl)
+      return false;
+
+    if (derivedDecl == baseDecl)
+      return true;
+
+    return derivedDecl->isDerivedFrom(baseDecl);
+  }
```
Comment (vgvassilev) [inline lib/Interpreter/InterOp.cpp:2206]:
```suggestion
  bool IsTypeDerivedFrom(SemaType *S, TCppType_t derived, TCppType_t base)
  {
    auto derivedType = clang::QualType::getFromOpaquePtr(derived);
    auto baseType = clang::QualType::getFromOpaquePtr(base);
    return S->isDerivedFrom(derivedType, baseType);
  }
```

I was suggesting something like that.

**Comment by Smit1603** [comment]
- @vgvassilev sry for misconception on my side
- I have modified code as mentioned by you
- I hope the module is correctly implemented now
- Pls review and let me know whether any further improvements are required or not

**Comment by vgvassilev** [comment]
> * @vgvassilev sry for misconception on my side
> 
>     * I have modified code as mentioned by you
> 
>     * I hope the module is correctly implemented now
> 
>     * Pls review and let me know whether any further improvements are required or not

Please write a good commit message and make this PR with a single commit.

**Comment by Smit1603** [comment]
- @vgvassilev and @sudo-panda I have squashed all commit as asked
- Pls review and let me know whether is it done or not

**Comment by vgvassilev** [comment]
As a commit message I can suggest: "Add an API checking if a type is derived from another."

**Comment by Smit1603** [comment]
- @vgvassilev I hope this works

File: include/clang/Interpreter/InterOp.h [hunk 9]
```diff
@@ -161,6 +161,8 @@ namespace InterOp {
 
   TCppType_t GetTypeFromScope(TCppScope_t klass);
 
+  bool IsTypeDerivedFrom(TCppSema_t sema,TCppType_t derived, TCppType_t base);
```
Comment (vgvassilev) [inline include/clang/Interpreter/InterOp.h:164]:
```suggestion
  /// Check if a C++ type derives from another.
  bool IsTypeDerivedFrom(TCppSema_t sema, TCppType_t derived, TCppType_t base);
```

File: lib/Interpreter/InterOp.cpp [hunk 10]
```diff
@@ -2184,5 +2184,15 @@ namespace InterOp {
     }
     return dims;
   }
+
+  bool IsTypeDerivedFrom(TCppSema_t sema,TCppType_t derived, TCppType_t base)
```
Comment (vgvassilev) [inline lib/Interpreter/InterOp.cpp:2188]:
```suggestion
  bool IsTypeDerivedFrom(TCppSema_t sema, TCppType_t derived, TCppType_t base)
```

**Review by vgvassilev — APPROVED** [review]
LGTM assuming the comments are addressed.

File: lib/Interpreter/InterOp.cpp (hunk 10 above)
Comment (Smit1603) [reply lib/Interpreter/InterOp.cpp]: Have performed this minor changes

**Pull request merged**